rfcs/text/0004-support-flow-in-monorepo.md

```diff
@@ -0,0 +1,58 @@
+
+# Background
+
+## Rush
+
+[`rush flow`](https://github.com/fusionjs/fusionjs/blob/master/common/config/rush/command-line.json#L29) currently runs as a "global" command, which essentially runs `flow check` at the root of the monorepo.  This is similar to how `yarn flow` runs `flow check` in the multi-repo setup for Fusion.js packages, but for only a single project.
+
+## Flow
+
+Flow supports a single [`.flowconfig`](https://flow.org/en/docs/config/) file.  This file serves two purposes.  Firstly, it is used to configure Flow for that specific project.  Secondly, it also defines the project root for Flow.  Because of this, it is [not intended to be used to construct a recursive configuration](https://github.com/facebook/flow/issues/238#issuecomment-72089643).  In other words, only the first ancestral `.flowconfig` file is used when running Flow.
+
+Flow also supports [library definition](https://flow.org/en/docs/libdefs/) (libdef) files which, by default, exist in the `<PROJECT_ROOT>/flow-typed/` directory.  These define type information, generally at the module level, for third-party dependencies. 
+
+# Problem
+
+Not surprisingly, a monorepo may cause issues as each project defines its own `.flowconfig` file, many of which are not consistent across repositories.
+
+For example, [`create-fusion-app`](https://github.com/fusionjs/fusionjs/blob/master/create-fusion-app/.flowconfig) defines a number of ignored directories:
+```
+[ignore]
+.*/templates/.*
+.*/test-artifacts/.*
+```
+
+Whereas [`fusion-tokens`](https://github.com/fusionjs/fusionjs/blob/master/fusion-tokens/.flowconfig) defines none.
+
+Furthermore, Flow resolves libdef files in the `<PROJECT_ROOT>/flow-typed/` directory by default.  Given that each project has its own `flow-typed` directory, and no project is at the "root" anymore, Flow fails to resolve any of these when running the `rush flow` command.  Unfortunately, Flow will not complain about missing type information, instead opting to resolve all untyped imports to `any`.  This may cause type coverage regressions.
+
+# Solution
+
+Align on a single `.flowconfig` file to live in the root of the monorepo.  While one exists today, it importantly misses many of the library definition (libdef) files defined by individual projects' configurations.
+
+We must ensure all relevant libdef files are included in the root configuration.  Any custom configuration in a single project must also be either supported across the entire monorepo, or removed.
+
+If multiple projects require a libdef for the same dependency, we ideally house only a single source of truth for that type information.  In this way, the monorepo solution cuts down on duplicated code and ensures consistency across projects.
```

Comment:
I thought part of the problem was also that we publish the libdefs along with any package that uses them. How would we handle publishing if the flow-typed directory exists outside the package?

Reply:
Good point!  In theory, we probably should not be publishing packages with libdef files.  Best practice would be that our packages are either entirely self-contained (i.e. no need for any definitions from a libdef file) or rely only upon open-sourced definitions that can be installed directly from [flow-typed](https://github.com/flow-typed/flow-typed).

Happy to bikeshed this as well!

Reply:
If we do get rid of this, we should look for an improvement in the process, like validating or automatically installing the correct libdefs.

Reply:
Yup, I think we'd probably want explore to install libdefs as part of scaffolding.